_posts/2019/2019-04-02-how-to-easily-manage-ssh-keys-part-2.md

```diff
@@ -0,0 +1,177 @@
+        date = v["f_text_c_1508155981380"].(string)
+      }
+      if v["f_text_c_1510835561457"] != nil {
+        label = v["f_text_c_1510835561457"].(string)
+      }
+      if v["f_dialog_c_1510835574647"].(map[string]interface{})["title"] != nil {
+        primary = v["f_dialog_c_1510835574647"].(map[string]interface{})["title"].(string)
+      }
+
+      id, err := strconv.Atoi(v["id"].(string))
+      if err != nil {
+        // put your error handling here
+      }
+      sshKeys = append(sshKeys, SSHKey{id, key, label, primary, date})
+    }
+  }
+```
+Dont be confused by lines like this
+
+```
+  v["f_dialog_c_1510835574647"].(map[string]interface{})["title"].(string)
+```
+the `.(map[string]interface{})` is a concept in go called [type assertions](https://tour.golang.org/methods/15){:target="_blank"} that just assigns a specific type to an interface.
+In our case we assign `map[string]interface{}` to it which is similar to an associative array where the key is a string and the value itself is also just `interface{}`.
+Then the index `title` is called and the `string` type is assigned to the interface.
+I say similar to arrays because go does not often make use of arrays, it has [slices](https://www.godesignpatterns.com/2014/05/arrays-vs-slices.html){:target="_blank"}.
+
+If you want to add a key to a user you can do this more or less the same way:
+```
+  keyData := struct {
+    SSHKey  string `json:"f_text_c_1506425983066"`
+    Date    string `json:"f_text_c_1508155981380"`
+    Label   string `json:"f_text_c_1510835561457"`
+    Primray string `json:"f_dialog_c_1510835574647"`
+    Desc    string `json:"description"`
+  }{key, date, label, primary, ""}
+
+  _, err = a.CreateCat(id, "C__CATG__CUSTOM_FIELDS_SSHKEYS", keyData)
+  if err != nil {
+    // put your error handling here
+  }
+```
+this adds a new ssh-key.
+
+We now know everyting to implement specific filters and combinations e.g:
+
+- selecting a group from our cmdb and getting all members and then getting all ssh-keys (and maybe filter them)
+- we could easly just add people to existing querys that already contain one or more groups
+- of course filtering the label, date, primary whatever is no problem
+
+Also our beloved friend `curl` the comandline data transfer tool can be used against our API.
+
+An example could look like this:
+
+```
+curl -X POST -H "Authorization: Basic $(echo 'user:password' | base64)" \
+ -d "[]group=some Team" \
+ -d "[]group=some other Team" \
+ -d "[]add=some additional user" \
+ -d "[]label=mylabel1" \
+ -d "[]label=mylabel2" \
+ -d "raw=true" \
+ https://ssh-api-url/key/export | jq -r '.["authorized_keys"]'
+```
+
+The `[]group` array is used to add groups to our query `[]add` to simply add person and `[]label` lets us filter for specific labels.
+I added the `raw` attribute to be able to get the keys as a data block instead of getting every key via a seperate JSON structure.
+
+which is just a difference between:
```

Comment:
What does "which" refer to? I'm missing the relation again.

Reply:
I deleted this one since its not nessesary to understand the basic concept. It was more like an addition